ipam/provider/aws.go

```diff
@@ -0,0 +1,256 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"time"
+
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/e2e-harness/pkg/framework"
+	awsclient "github.com/giantswarm/e2eclients/aws"
+	"github.com/giantswarm/e2etemplates/pkg/chartvalues"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type AWSConfig struct {
+	AWSClient     *awsclient.Client
+	HostFramework *framework.Host
+	Logger        micrologger.Logger
+
+	ChartValuesConfig
+}
+
+type AWS struct {
+	awsClient     *awsclient.Client
+	hostFramework *framework.Host
+	logger        micrologger.Logger
+
+	chartValuesConfig ChartValuesConfig
+	hostClusterName   string
+}
+
+type ChartValuesConfig struct {
+	AWSAPIHostedZoneGuest     string
+	AWSIngressHostedZoneGuest string
+	AWSRegion                 string
+	AWSRouteTable0            string
+	AWSRouteTable1            string
+	AWSVPCPeerID              string
+	ClusterName               string
+	CommonDomain              string
+	SSHPublicKey              string
+	VersionBundleVersion      string
+}
+
+func NewAWS(config AWSConfig) (*AWS, error) {
+	if config.AWSClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AWSClient must not be empty", config)
+	}
+	if config.HostFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostFramework must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	a := &AWS{
+		awsClient:     config.AWSClient,
+		hostFramework: config.HostFramework,
+		logger:        config.Logger,
+
+		chartValuesConfig: ChartValuesConfig{
+			AWSAPIHostedZoneGuest:     getValue(config.AWSAPIHostedZoneGuest, "AWS_API_HOSTED_ZONE_GUEST"),
+			AWSIngressHostedZoneGuest: getValue(config.AWSIngressHostedZoneGuest, "AWS_INGRESS_HOSTED_ZONE_GUEST"),
+			AWSRegion:                 getValue(config.AWSRegion, "AWS_REGION"),
+			AWSRouteTable0:            getValue(config.AWSRouteTable0, "AWS_ROUTE_TABLE_0"),
+			AWSRouteTable1:            getValue(config.AWSRouteTable1, "AWS_ROUTE_TABLE_1"),
+			CommonDomain:              getValue(config.CommonDomain, "COMMON_DOMAIN"),
+			SSHPublicKey:              getValue(config.SSHPublicKey, "IDRSA_PUB"),
+			VersionBundleVersion:      getValue(config.VersionBundleVersion, "VERSION_BUNDLE_VERSION"),
+		},
+		hostClusterName: getValue(config.ClusterName, "CLUSTER_NAME"),
+	}
```

Comment:
I think it would be good to do all the env stuff in separate `env.go`.

Reply:
Roger. I'll refactor this.